File: plugin/src/knowledge/providers/openlibrary.ts
import { requestUrl } from "obsidian";
import type { KnowledgeProvider, LookupOptions, LookupResult, KnowledgeEntry } from "../types";

const OPENLIBRARY_API = "https://openlibrary.org";

interface OpenLibrarySearchResult {
  docs?: Array<{
    key: string;
    title: string;
    author_name?: string[];
    first_publish_year?: number;
    isbn?: string[];
    subject?: string[];
    cover_i?: number;
  }>;
  numFound?: number;
}

interface OpenLibraryAuthorResult {
  docs?: Array<{
    key: string;
    name: string;
    birth_date?: string;
    death_date?: string;
    top_work?: string;
    work_count?: number;
  }>;
}

interface OpenLibraryWorkResult {
  title?: string;
  description?: string | { value: string };
  subjects?: string[];
  authors?: Array<{ author: { key: string } }>;
}

export class OpenLibraryProvider implements KnowledgeProvider {
  readonly name = "openlibrary";

  async isAvailable(): Promise<boolean> {
    try {
      const response = await requestUrl({
        url: `${OPENLIBRARY_API}/search.json?q=test&limit=1`,
        method: "GET",
        throw: false,
      });
      return response.status === 200;
    } catch {
      return false;
    }
  }

  async lookup(query: string, options?: LookupOptions): Promise<LookupResult> {
    const maxResults = options?.maxResults ?? 5;

    try {
      const entries: KnowledgeEntry[] = [];

      // Search books first
      const bookResults = await this.searchBooks(query, maxResults);
      entries.push(...bookResults);

      // If few results, also search authors
      if (entries.length < maxResults) {
        const authorResults = await this.searchAuthors(query, maxResults - entries.length);
        entries.push(...authorResults);
      }

      return {
        success: true,
        provider: this.name,
        entries,
      };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Unknown error";
      return {
        success: false,
        provider: this.name,
        entries: [],
        error: `OpenLibrary lookup failed: ${message}`,
      };
    }
  }

  private async searchBooks(query: string, limit: number): Promise<KnowledgeEntry[]> {
    const params = new URLSearchParams({
      q: query,
      limit: String(limit),
      fields: "key,title,author_name,first_publish_year,isbn,subject,cover_i",
    });

    const response = await requestUrl({
      url: `${OPENLIBRARY_API}/search.json?${params}`,
      method: "GET",
      throw: false,
    });

    if (response.status !== 200) {
      return [];
    }

    const data = response.json as OpenLibrarySearchResult;

    if (!data.docs || !Array.isArray(data.docs)) {
      return [];
    }

    return data.docs.map((book) => {
      const authors = book.author_name?.join(", ") ?? "Unknown author";
      const year = book.first_publish_year ? ` (${book.first_publish_year})` : "";

      return {
        title: book.title,
        summary: `${authors}${year}`,
        url: `https://openlibrary.org${book.key}`,
        source: "openlibrary",
        metadata: {
          type: "book",
          authors: book.author_name,
          year: book.first_publish_year,
          isbn: book.isbn?.[0],
          subjects: book.subject?.slice(0, 5),
          coverId: book.cover_i,
        },
      };
    });
  }

  private async searchAuthors(query: string, limit: number): Promise<KnowledgeEntry[]> {
    const params = new URLSearchParams({
      q: query,
      limit: String(limit),
    });

    const response = await requestUrl({
      url: `${OPENLIBRARY_API}/search/authors.json?${params}`,
      method: "GET",
      throw: false,
    });

    if (response.status !== 200) {
      return [];
    }

    const data = response.json as OpenLibraryAuthorResult;

    if (!data.docs || !Array.isArray(data.docs)) {
      return [];
    }

    return data.docs.map((author) => {
      const years =
        author.birth_date || author.death_date
          ? ` (${author.birth_date ?? "?"} - ${author.death_date ?? ""})`
          : "";
      const works = author.work_count ? `, ${author.work_count} works` : "";
      const topWork = author.top_work ? `. Notable: "${author.top_work}"` : "";

      return {
        title: author.name,
        summary: `Author${years}${works}${topWork}`,
        url: `https://openlibrary.org/authors/${author.key.replace("/authors/", "")}`,
        source: "openlibrary",
        metadata: {
          type: "author",
          birthDate: author.birth_date,
          deathDate: author.death_date,
          workCount: author.work_count,
          topWork: author.top_work,
        },
      };
    });
  }

  async getWorkDetails(workKey: string): Promise<KnowledgeEntry | null> {
    const response = await requestUrl({
      url: `${OPENLIBRARY_API}${workKey}.json`,
      method: "GET",
      throw: false,
    });

    if (response.status !== 200) {
      return null;
    }

    const data = response.json as OpenLibraryWorkResult;

    const description =
      typeof data.description === "string"
        ? data.description
        : data.description?.value ?? "";

    return {
      title: data.title ?? "Unknown",
      summary: description.slice(0, 500),
      url: `https://openlibrary.org${workKey}`,
      source: "openlibrary",
      metadata: {
        type: "work",
        subjects: data.subjects?.slice(0, 10),
      },
    };
  }
}

File: crates/mcp/src/knowledge/openlibrary.rs
use reqwest::blocking::Client;
use serde::Deserialize;
use std::collections::HashMap;

use super::{KnowledgeEntry, KnowledgeProvider, LookupOptions, LookupResult};

const OPENLIBRARY_API: &str = "https://openlibrary.org";

pub struct OpenLibraryProvider {
    client: Client,
}

#[derive(Debug, Deserialize)]
struct BookSearchResponse {
    docs: Option<Vec<BookDoc>>,
}

#[derive(Debug, Deserialize)]
struct BookDoc {
    key: String,
    title: String,
    author_name: Option<Vec<String>>,
    first_publish_year: Option<i32>,
    isbn: Option<Vec<String>>,
    subject: Option<Vec<String>>,
    cover_i: Option<i64>,
}

#[derive(Debug, Deserialize)]
struct AuthorSearchResponse {
    docs: Option<Vec<AuthorDoc>>,
}

#[derive(Debug, Deserialize)]
struct AuthorDoc {
    key: String,
    name: String,
    birth_date: Option<String>,
    death_date: Option<String>,
    top_work: Option<String>,
    work_count: Option<i32>,
}

impl OpenLibraryProvider {
    pub fn new() -> Self {
        Self {
            client: Client::builder()
                .user_agent("vault-tree-mcp/0.1 (https://github.com/Hydepwns/vault-tree)")
                .build()
                .unwrap_or_else(|_| Client::new()),
        }
    }

    fn search_books(&self, query: &str, limit: usize) -> Result<Vec<KnowledgeEntry>, String> {
        let url = format!(
            "{}/search.json?q={}&limit={}&fields=key,title,author_name,first_publish_year,isbn,subject,cover_i",
            OPENLIBRARY_API,
            urlencoding::encode(query),
            limit
        );

        let response = self.client.get(&url).send().map_err(|e| e.to_string())?;

        if !response.status().is_success() {
            return Ok(Vec::new());
        }

        let data: BookSearchResponse = response.json().map_err(|e| e.to_string())?;

        Ok(data
            .docs
            .unwrap_or_default()
            .into_iter()
            .map(|book| {
                let authors = book.author_name.as_ref().map(|a| a.join(", ")).unwrap_or_else(|| "Unknown author".to_string());
                let year = book.first_publish_year.map(|y| format!(" ({})", y)).unwrap_or_default();

                let mut metadata = HashMap::new();
                metadata.insert("type".to_string(), serde_json::json!("book"));
                if let Some(ref author_list) = book.author_name {
                    metadata.insert("authors".to_string(), serde_json::json!(author_list));
                }
                if let Some(y) = book.first_publish_year {
                    metadata.insert("year".to_string(), serde_json::json!(y));
                }
                if let Some(ref isbns) = book.isbn {
                    if let Some(first_isbn) = isbns.first() {
                        metadata.insert("isbn".to_string(), serde_json::json!(first_isbn));
                    }
                }
                if let Some(ref subjects) = book.subject {
                    let limited: Vec<_> = subjects.iter().take(5).cloned().collect();
                    metadata.insert("subjects".to_string(), serde_json::json!(limited));
                }
                if let Some(cover_id) = book.cover_i {
                    metadata.insert("coverId".to_string(), serde_json::json!(cover_id));
                }

                KnowledgeEntry {
                    title: book.title,
                    summary: format!("{}{}", authors, year),
                    url: Some(format!("https://openlibrary.org{}", book.key)),
                    source: "openlibrary".to_string(),
                    metadata: Some(metadata),
                }
            })
            .collect())
    }

    fn search_authors(&self, query: &str, limit: usize) -> Result<Vec<KnowledgeEntry>, String> {
        let url = format!(
            "{}/search/authors.json?q={}&limit={}",
            OPENLIBRARY_API,
            urlencoding::encode(query),
            limit
        );

        let response = self.client.get(&url).send().map_err(|e| e.to_string())?;

        if !response.status().is_success() {
            return Ok(Vec::new());
        }

        let data: AuthorSearchResponse = response.json().map_err(|e| e.to_string())?;

        Ok(data
            .docs
            .unwrap_or_default()
            .into_iter()
            .map(|author| {
                let years = match (&author.birth_date, &author.death_date) {
                    (Some(b), Some(d)) => format!(" ({} - {})", b, d),
                    (Some(b), None) => format!(" ({} - )", b),
                    (None, Some(d)) => format!(" (? - {})", d),
                    (None, None) => String::new(),
                };
                let works = author.work_count.map(|c| format!(", {} works", c)).unwrap_or_default();
                let top_work = author.top_work.as_ref().map(|w| format!(". Notable: \"{}\"", w)).unwrap_or_default();

                let mut metadata = HashMap::new();
                metadata.insert("type".to_string(), serde_json::json!("author"));
                if let Some(ref b) = author.birth_date {
                    metadata.insert("birthDate".to_string(), serde_json::json!(b));
                }
                if let Some(ref d) = author.death_date {
                    metadata.insert("deathDate".to_string(), serde_json::json!(d));
                }
                if let Some(c) = author.work_count {
                    metadata.insert("workCount".to_string(), serde_json::json!(c));
                }
                if let Some(ref w) = author.top_work {
                    metadata.insert("topWork".to_string(), serde_json::json!(w));
                }

                let author_key = author.key.trim_start_matches("/authors/");

                KnowledgeEntry {
                    title: author.name,
                    summary: format!("Author{}{}{}", years, works, top_work),
                    url: Some(format!("https://openlibrary.org/authors/{}", author_key)),
                    source: "openlibrary".to_string(),
                    metadata: Some(metadata),
                }
            })
            .collect())
    }
}

impl Default for OpenLibraryProvider {
    fn default() -> Self {
        Self::new()
    }
}

impl KnowledgeProvider for OpenLibraryProvider {
    fn name(&self) -> &'static str {
        "openlibrary"
    }

    fn is_available(&self) -> bool {
        self.client
            .get(format!("{}/search.json?q=test&limit=1", OPENLIBRARY_API))
            .send()
            .map(|r| r.status().is_success())
            .unwrap_or(false)
    }

    fn lookup(&self, query: &str, options: &LookupOptions) -> LookupResult {
        let limit = options.max_results.unwrap_or(5);

        let mut entries = match self.search_books(query, limit) {
            Ok(e) => e,
            Err(e) => return LookupResult::error(self.name(), e),
        };

        if entries.len() < limit {
            let remaining = limit - entries.len();
            match self.search_authors(query, remaining) {
                Ok(authors) => entries.extend(authors),
                Err(e) => return LookupResult::error(self.name(), e),
            }
        }

        LookupResult::success(self.name(), entries)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    #[ignore] // Requires network
    fn openlibrary_lookup() {
        let provider = OpenLibraryProvider::new();
        let result = provider.lookup("dune frank herbert", &LookupOptions::default());
        assert!(result.success);
        assert!(!result.entries.is_empty());
    }
}
